crates/core/src/fiat_shamir/hasher_challenger.rs

```diff
@@ -21,75 +22,141 @@ pub struct HasherSampler<H: Digest> {
 /// Challenger type which implements `[BufMut]` that has similar functionality as `[CanObserve]`
 #[derive(Debug, Default)]
 pub struct HasherObserver<H: Digest + BlockSizeUser> {
+	observation_count: usize,
 	index: usize,
 	buffer: Block<H>,
 	hasher: H,
 }
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Interaction {
+	Observe(usize),
+	Sample(usize),
+}
+
+#[derive(Debug)]
+pub enum ObserverOrSampler<H: Digest + BlockSizeUser> {
+	Observer(HasherObserver<H>),
+	Sampler(HasherSampler<H>),
+}
+
+impl<H: Digest + BlockSizeUser + Default> Default for ObserverOrSampler<H> {
+	fn default() -> Self {
+		Self::Sampler(HasherSampler::default())
+	}
+}
+
 /// Challenger interface over hashes that implement `[Digest]` trait,
 ///
 /// This challenger works over bytes instead of Field elements
 #[derive(Debug)]
-pub enum HasherChallenger<H: Digest + BlockSizeUser> {
-	Observer(HasherObserver<H>),
-	Sampler(HasherSampler<H>),
+pub struct HasherChallenger<H: Digest + BlockSizeUser> {
+	interactions: Vec<Interaction>,
+	observer_or_sampler: ObserverOrSampler<H>,
 }
 
 impl<H> Default for HasherChallenger<H>
 where
-	H: Digest + BlockSizeUser,
+	H: Digest + BlockSizeUser + FixedOutputReset,
 {
 	fn default() -> Self {
-		Self::Observer(HasherObserver {
-			hasher: H::new(),
-			index: 0,
-			buffer: Block::<H>::default(),
-		})
+		Self::new(H::digest([]))
+	}
+}
+
+/// Initialize the `HasherChallenger` with an `initial_digest` that reflects both the statement and the constraint system,
+/// e.g. had by hashing the hash of the statement and the hash of the serialized constraint system.
+impl<H> HasherChallenger<H>
+where
+	H: Digest + BlockSizeUser,
+{
+	fn new(initial_digest: Output<H>) -> Self {
+		let mut hasher = H::new();
+		Digest::update(&mut hasher, &initial_digest);
+
+		Self {
+			interactions: Vec::new(),
+			observer_or_sampler: ObserverOrSampler::Sampler(HasherSampler {
+				sample_count: 0,
+				hasher,
+				index: 0,
+				buffer: initial_digest,
```

Comment:
I'm not sure, but do we really need to use initial_digest as a buffer after adding it to the hasher?